<!DOCTYPE html>
<html>
<head>
<title>Maftools_practical.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h1 id="generating-oncoplots-using-maftools">Generating oncoplots using maftools</h1>
<p>Oncoplots (also called waterfall plots) are useful plots that help us visualize frequenty affected genes in a cohort. This type of plots can also display information about the number and type of mutations, and can include annotations about the characteristics of the cohort, such as clinical data. The main structure of this plot is to show in each column a sample and in each row a gene. Swhowing mutations for the displayed genes for each sample with a color code.</p>
<p>There are several tools that can be used to generate this type of plots. Today we will be using the R package maftools (Mayakonda et al., 2018).</p>
<p>We will now be using the TCGA-BLCA data in maf format to generate our own oncoplots.</p>
<h2 id="reading-maf-files-into-maftools">Reading MAF files into maftools</h2>
<p>First we need to set the directory were we will be working.</p>
<pre class="hljs"><code><div>setwd(<span class="hljs-string">"/home/manager/data/Driver_detection-20220823T163817Z-001/Driver_detection/"</span>)
</div></code></pre>
<p>Now, we'll call the maftools package.</p>
<pre class="hljs"><code><div><span class="hljs-keyword">library</span>(maftools)
</div></code></pre>
<p>Today's data is already merged in one MAF file. To read any MAF or MAF like file using maftools use the following command.</p>
<pre class="hljs"><code><div>BLCA = read.maf(maf = <span class="hljs-string">"TCGA-BLCA.maf"</span>)
</div></code></pre>
<p>With this we have read the MAF file and stored the data in the object BLCA.</p>
<p>While reading the MAF file you can notice that in the console some information about the data is being displayed. This includes information about the number of silent variants in the data, genes with possible flags and, in our case, it states that we have not included clinical data.</p>
<h2 id="creating-an-oncoplot">Creating an oncoplot</h2>
<p>Now that we have all the data from the cohort stored in one object, we will plot a basic oconplot.</p>
<pre class="hljs"><code><div>oncoplot(maf= BLCA)
</div></code></pre>
<p><img src="plots/first_oncoplot.png" alt="First_oncoplot"></p>
<p>As you can see, with this simple command you can get an oncoplot displaying 20 genes.</p>
<p>With the default parameters the oncoplot will also show the percentage of samples in the cohort that have alterations in the shown genes, as well as the type of mutations affecting the gene through a code of colors.</p>
<p>From this oncoplot, how do you think maftools selects the genes to show in the plot?</p>
<p>Compare these genes with the data you obtained from dndscv.</p>
<p>We can change the number of genes displayed with the option &quot;top&quot;.</p>
<p>For example, if we want to show 30 genes we can type the following.</p>
<pre class="hljs"><code><div>oncoplot(maf=BLCA, top=<span class="hljs-number">30</span>)
</div></code></pre>
<p><img src="plots/30genes_oncoplot.png" alt="First_oncoplot"></p>
<p>Or if we want to only display 10 genes we just type:</p>
<pre class="hljs"><code><div>oncoplot(maf=BLCA, top=<span class="hljs-number">10</span>)
</div></code></pre>
<p><img src="plots/10genes_oncoplot.png" alt="First_oncoplot"></p>
<p>As you may have noticed, the genes shown are selected only considering the number of mutations affecting them, and the size or composition of these genes is not taken into account.</p>
<p>For example the TTN gene (polymorphic gene) is a very large gene. Because of its size it is expected to have a higher amount of mutations affecting it. Therefore it may not be relevant to this type of cancer. If you return to the information displayed when reading the maf file you can also see that the TTN gene was one of the genes with possible flags.</p>
<p>You can customize you oncoplot taking this type of considerations. For example lets try to exclude the TTN gene from our oncoplot.</p>
<pre class="hljs"><code><div>oncoplot(BLCA, genesToIgnore = c(<span class="hljs-string">"TTN"</span>))
</div></code></pre>
<p><img src="plots/TTN_OUT.png" alt="First_oncoplot"></p>
<p>Now that we have used dndscv to identify driver genes, we can also select those genes to generate a specific oncoplot with them.</p>
<p>Let's try to generate an oncoplot with 5 significant genes we identified with dndscv.</p>
<pre class="hljs"><code><div>oncoplot(BLCA, genes = c(<span class="hljs-string">"TP53"</span>,<span class="hljs-string">"PIK3CA"</span>,<span class="hljs-string">"ARID1A"</span>,<span class="hljs-string">"KMT2D"</span>,<span class="hljs-string">"RB1"</span>))
</div></code></pre>
<p><img src="plots/5genes_oncoplot.png" alt="First_oncoplot"></p>
<p>As mentioned at the beginning, the plots are also known as waterfall plots due to their appearance. What do you think causes the plot to look like this? (Compare samples and genes) Do you see any patterns?</p>
<p>To make this comparison easier let's create an oncoplot only for the genes KMT2D and KDM6A.</p>
<pre class="hljs"><code><div>oncoplot(BLCA, genes = c(<span class="hljs-string">"KMT2D"</span>,<span class="hljs-string">"KDM6A"</span>))
</div></code></pre>
<p><img src="plots/MUTUALLY_EXCLUSIVE.png" alt="First_oncoplot"></p>
<p>What can you see here? (Remember each &quot;column&quot; is one sample)</p>
<p>The KDM6A and KMT2D genes have shown frequently to be mutually exclusive in bladder cancer (Sangster et al., 2022). In our cohort you can see that most samples that have a mutation in one of these genes do not have a mutation in the other. Just a few samples have mutations in both genes. This type of relationship between these genes can indicate that both genes play similar roles in this type of cancer.</p>
<p>Oncoplots can help us detect certain relatioships between affected genes, as shown above.</p>
<h1 id="adding-annotations">Adding annotations</h1>
<p>We can also add other annotations to our oncoplots, such as transitions and transvertions. We can add this information like this:</p>
<pre class="hljs"><code><div>oncoplot(BLCA, genes = c(<span class="hljs-string">"TP53"</span>,<span class="hljs-string">"PIK3CA"</span>,<span class="hljs-string">"ARID1A"</span>,<span class="hljs-string">"KMT2D"</span>,<span class="hljs-string">"RB1"</span>), draw_titv = <span class="hljs-literal">TRUE</span>)
</div></code></pre>
<p>Being able to visualize transitions and transvertions can help us indetify certain patterns that could be related to a specific source of DNA damage.</p>
<p>We can also use the annotations of the variant classification that is displayed by default to identify oncogenes and tumor suppressors. What type of mutations do you expect on oncogenes? And what type do you expect on tumor suppresors?</p>
<p>Depending on your cohort, it can also be useful to have the IDs of the samples on the plot. You can add them with the option &quot;showTumorSampleBarcodes&quot;.</p>
<pre class="hljs"><code><div>oncoplot(BLCA, genes = c(<span class="hljs-string">"TP53"</span>,<span class="hljs-string">"PIK3CA"</span>,<span class="hljs-string">"ARID1A"</span>,<span class="hljs-string">"KMT2D"</span>,<span class="hljs-string">"RB1"</span>), draw_titv = <span class="hljs-literal">TRUE</span>, showTumorSampleBarcodes=<span class="hljs-literal">TRUE</span>)
</div></code></pre>
<p><img src="plots/sample_names.png" alt="First_oncoplot"></p>
<h2 id="adding-pathways">Adding pathways</h2>
<p>Maftools has the option to plot affected pathways and the genes corresponding to those pathways. By using the option pathways = 'auto', maftools will display the three most affected pathways.</p>
<pre class="hljs"><code><div>oncoplot(maf=BLCA, pathways= <span class="hljs-string">'auto'</span>, gene_mar =<span class="hljs-number">5</span>, fontSize =<span class="hljs-number">0.3</span>)
</div></code></pre>
<p><img src="plots/pathways_oncoplot.png" alt="First_oncoplot"></p>
<p>The option &quot;gene_mar&quot; in the previous command changes the margin width in gene names, and &quot;fontSize&quot; changes the size of the gene names to improve readability in the plot.</p>
<h2 id="other-maftools-plots">Other maftools plots</h2>
<p>Maftools can generate other useful plots. For example it can plot a summary of all variants, showing variants per sample, variant type and classification.</p>
<p>Use the next command to get the summary plot.</p>
<pre class="hljs"><code><div>plotmafSummary(maf=BLCA, addStat = <span class="hljs-string">'median'</span>, dashboard = <span class="hljs-literal">TRUE</span>)
</div></code></pre>
<p><img src="plots/maf_summary.png" alt="First_oncoplot"></p>
<p>We can also generate a specific plot for transitions and transvertions. To do so, we first get the transition and transvertion data with &quot;titv&quot; and then plot it with &quot;plotTiTv&quot;.</p>
<pre class="hljs"><code><div>BLCA_titv = titv(maf = BLCA, plot=<span class="hljs-literal">FALSE</span>)
plotTiTv(res=BLCA_titv)
</div></code></pre>
<p><img src="plots/titv.png" alt="First_oncoplot"></p>
<p>Another plot that we can create is a lollipop plot. This plot shows the positions and amino acid changes that are caused in the protein by the mutations present. You can create this type of plot for your protein of interest with the following command:</p>
<pre class="hljs"><code><div><span class="hljs-comment">#We will use TP53 as an example</span>
lollipopPlot(maf=BLCA, gene=<span class="hljs-string">"TP53"</span>, showMutationRate=<span class="hljs-literal">TRUE</span>)
</div></code></pre>
<p><img src="plots/lollipop1.png" alt="First_oncoplot"></p>
<p>You can label the changes by using the option &quot;labelPos&quot; and stating the position of the change you would like to annotate.</p>
<p>Here's an example.</p>
<pre class="hljs"><code><div><span class="hljs-comment">#Here, we will label teh change in the position 20.</span>
lollipopPlot(maf=BLCA, gene=<span class="hljs-string">"TP53"</span>, showMutationRate=<span class="hljs-literal">TRUE</span>, labelPos=<span class="hljs-number">20</span>)
</div></code></pre>
<p><img src="plots/lollipop2.png" alt="First_oncoplot"></p>
<p>You can label as many positions as you want.</p>
<pre class="hljs"><code><div>lollipopPlot(maf=BLCA, gene=<span class="hljs-string">"TP53"</span>, showMutationRate=<span class="hljs-literal">TRUE</span>, labelPos=c(<span class="hljs-number">20</span>,<span class="hljs-number">337</span>))
</div></code></pre>
<p><img src="plots/lollipop4.png" alt="First_oncoplot"></p>
<p>An important thing to consider for these plots is that to be able to create them, you need the information of the change of amino acid to be included in your maf file. As well, the name of the column for this information can vary between maf files. By default maftools will look for the name &quot;AAChange&quot;. If the column has another name, you should use the option &quot;AACol&quot; to state the name used in your maf file.</p>
<h2 id="when-you-have-per-sample-mafs">When you have per sample mafs</h2>
<p>This is not the case for this practical, but it is common to have per sample mafs. In these cases you can merge them together using maftools and create your plots as above. If you need to merge several maf files you can use the following commands.</p>
<pre class="hljs"><code><div>mafs = list.files(path = <span class="hljs-string">"/directory/maf_files/"</span>, pattern = <span class="hljs-string">"*.\\.maf$"</span>, full.names =<span class="hljs-literal">TRUE</span>)
all_samples = merge_mafs(mafs = mafs)

</div></code></pre>
<p>In this command you must indicate the path to the directory where you have your files. The names will be stored in &quot;mafs&quot; and you can then use this with &quot;merge_mafs&quot; option of maftools. In this example the final merged maf is stored in &quot;all_samples&quot;.</p>
<p>References:</p>
<p>Mayakonda A, Lin D, Assenov Y, Plass C, Koeffler PH (2018). Maftools: efficient and comprehensive analysis of somatic variants in cancer. Genome Research. doi: 10.1101/gr.239244.118.</p>
<p>Sangster, A. G., Gooding, R. J., Garven, A., Ghaedi, H.,
Berman, D. M., &amp; Davey, S. K. (2022). Mutually exclusive mutation profiles define functionally related genes in muscle invasive bladder cancer. PloS one, 17(1), e0259992. https://doi.org/10.1371/journal.pone.0259992</p>

</body>
</html>
